# DATA620: Assignment 7

## Eric Lehmphul

In [1]:
import pandas as pd

# Introduction

Locations around the world can experience vastly different levels of weather. The particular dataset that I am using for Assignment 7 relates to sunshine duration. The dataset provides the number of hours of sunshine in a month for a given city. I chose this dataset because I was curious to how location can affect exposure to sunlight. I also would like to answer the following question: Do any cities experience an average of 16 or more hours of sunight daily across an entire year? I found the dataset on Kaggle.com at the following link: https://www.kaggle.com/datasets/prasertk/sunshine-duration-by-city

# Data Exploration
Import your dataset into your .ipynb, create dataframes, and explore your data.  

Include: 

* Summary statistics means, medians, quartiles, 
* Missing value information
* Any other relevant information about the dataset.  



### Read in Data

With all data projects, the first task is to read the data into memory and conduct some data exploration techniques.

In [2]:
sun = pd.read_csv('https://raw.githubusercontent.com/SaneSky109/DATA602/main/Assignments/Assignment7/Data/sunshine%20hours%20by%20city%20dirty.csv')

sun.drop(0, axis = 0, inplace = True) # drop row that set all data to string / object 
sun.head()

Location         City Month1 Month2 Month3 Month4 Month5 Month6 Month7  \
1  Afghanistan        Kabul  177.2  178.6  204.5  232.5  310.3  353.4  356.8   
2      Albania       Tirana    124    125    165    191    263    298    354   
3          NaN      Algiers    NaN    NaN    202    258    319    318    350   
4      Algeria  Tamanrasset  297.6  275.5  322.4    327  328.6    306  356.5   
5       Angola       Luanda    219    208    213    199    233    223    175   

  Month8 Month9 Month10 Month11 Month12   Total  
1  339.7  303.9   282.6   253.2   182.4  3175.1  
2    327    264     218     127      88    2544  
3    319    237     229     165     136    2847  
4  331.7    288     310     285   272.8    3686  
5    150    145     164     199     212    2341

### Look at Data Types from "Unclean" File

All data in this file appears to be stored as a string / object. This will not work moving forward with the data, as most of the data is numeric. The data types will need to be adjusted.

In [3]:
sun.dtypes

Location    object
City        object
Month1      object
Month2      object
Month3      object
Month4      object
Month5      object
Month6      object
Month7      object
Month8      object
Month9      object
Month10     object
Month11     object
Month12     object
Total       object
dtype: object

### Summary Statistics on "Unclean" File

Looking at this describe function output I notice 2 things:
* The data must be converted to see the necessary summary statistics (means, medians, quartiles)
* There are potential row duplicates that need to be investigated

In [4]:
sun.describe()

Location    City Month1 Month2 Month3 Month4 Month5 Month6  \
count             392     393    392    392    392    394    394    393   
unique            140     380    295    308    279    253    277    281   
top     United States  Harare    217  190.4  232.5    180    279    264   
freq               54       6      9      6     12      8      7      9   

       Month7 Month8 Month9 Month10 Month11 Month12   Total  
count     393    394    393     393     394     394     392  
unique    312    290    265     287     278     291     370  
top       279  300.7    294   285.2     231   198.4  3010.9  
freq       10      7      8       7       8       7       6

### Correct Data Types

The `.astype()` function can be used to update the data types of columns in the dataframe. I inputed the column and data type as a dictionary to input all updates in one function.

In [5]:
sunshine = sun.astype({'Month1':'float64',
                 'Month2':'float64',
                 'Month3':'float64',
                 'Month4':'float64',
                 'Month5':'float64',
                'Month6':'float64',
                'Month7':'float64',
                'Month8':'float64',
                'Month9':'float64',
                'Month10':'float64',
                'Month11':'float64',
                'Month12':'float64',
                'Total':'float64'})

sunshine.dtypes

Location     object
City         object
Month1      float64
Month2      float64
Month3      float64
Month4      float64
Month5      float64
Month6      float64
Month7      float64
Month8      float64
Month9      float64
Month10     float64
Month11     float64
Month12     float64
Total       float64
dtype: object

### Summary Statistics with Correct Data Types

Some notable observations from this dataset:
* There are missing values that need to be accounted for (counts are different)
* Month 1 (January) has a city somewhere that recieves 0 hours of sunlight 
* Month 7 (July) has the highest mean and median hours of sunlight and Month 12 (December) has the lowest mean and median hours of sunlight

#### Numeric Data

In [6]:
sunshine.describe()

Month1      Month2      Month3      Month4      Month5      Month6  \
count  392.000000  392.000000  392.000000  394.000000  394.000000  393.000000   
mean   180.709184  178.167730  207.780867  218.955330  241.550508  236.652672   
std     80.393666   64.286113   58.171712   52.712311   61.297374   73.085888   
min      0.000000   22.600000   47.000000   65.100000   68.200000   45.000000   
25%    124.000000  133.675000  164.300000  183.250000  197.125000  177.000000   
50%    182.450000  178.700000  210.800000  220.000000  248.900000  251.000000   
75%    245.050000  228.800000  251.325000  252.000000  284.000000  288.000000   
max    362.700000  314.900000  335.500000  365.500000  407.400000  419.500000   

           Month7      Month8      Month9     Month10     Month11     Month12  \
count  393.000000  394.000000  393.000000  393.000000  394.000000  394.000000   
mean   243.857506  236.411675  213.258524  207.183969  182.011929  173.250254   
std     80.743290   73.387831   62.839337   67.176687   76.927179   82.680421   
min     28.600000   32.300000   37.300000   24.800000    0.000000    0.000000   
25%    181.000000  186.000000  165.100000  156.000000  127.100000  118.350000   
50%    251.000000  242.000000  218.800000  211.000000  184.450000  171.500000   
75%    307.200000  295.000000  261.000000  263.000000  245.700000  238.025000   
max    440.200000  406.900000  347.800000  356.500000  363.000000  367.000000   

             Total  
count   392.000000  
mean   2517.701658  
std     620.689167  
min     840.000000  
25%    2041.900000  
50%    2507.700000  
75%    3002.500000  
max    4015.300000

#### Non-Numeric Data

In [7]:
sunshine.describe(exclude='float64')

Location    City
count             392     393
unique            140     380
top     United States  Harare
freq               54       6

# Data Wrangling
Create a subset of your original data and perform the following.  

1. Modify multiple column names.

2. Look at the structure of your data – are any variables improperly coded? Such as strings or characters? Convert to correct structure if needed.

3. Fix missing and invalid values in data.

4. Create new columns based on existing columns or calculations.

5. Drop column(s) from your dataset.

6. Drop a row(s) from your dataset. (drop duplicates)

7. Sort your data based on multiple variables. 

8. Filter your data based on some condition. 

9. Convert all the string values to upper or lower cases in one column.

10. Check whether numeric values are present in a given column of your dataframe.

11. Group your dataset by one column, and get the mean, min, and max values by group. 
  * Groupby()
  * agg() or .apply()

12. Group your dataset by two columns and then sort the aggregated results within the groups. 

**You are free (and should) to add on to these questions.  Please clearly indicate in your assignment your answers to these questions.**

## 1. Modify multiple column names.

Columns can be renamed using the rename function. I inputted a dictionary of all the columns that needed to be renamed. I outputed both the original column names and the new column names.

In [8]:
# Original Column Names
print("Original Column Names: \n", sunshine.columns)

# Update Column Names
sunshine = sunshine.rename(columns = {'Location':'Country',
                'Month1':'Jan',
                'Month2':'Feb',
                'Month3':'Mar',
                'Month4':'Apr',
                'Month5':'May',
                'Month6':'Jun',
                'Month7':'Jul',
                'Month8':'Aug',
                'Month9':'Sep',
                'Month10':'Oct',
                'Month11':'Nov',
                'Month12':'Dec',
                'Total':'Yearly'})

print("\nNew Column Names: \n", sunshine.columns)

Original Column Names: 
 Index(['Location', 'City', 'Month1', 'Month2', 'Month3', 'Month4', 'Month5',
       'Month6', 'Month7', 'Month8', 'Month9', 'Month10', 'Month11', 'Month12',
       'Total'],
      dtype='object')

New Column Names: 
 Index(['Country', 'City', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul',
       'Aug', 'Sep', 'Oct', 'Nov', 'Dec', 'Yearly'],
      dtype='object')


## 2. Look at the structure of your data – are any variables improperly coded? Such as strings or characters? Convert to correct structure if needed.

The original file considered everything as a string object. Most of the columns contain numeric data which warrants adjustments. As shown in the Data Exploration section above, the `.astype()` function can be used to update the data types of columns in the dataframe.

In [9]:
print("Original Data Types: \n", sun.dtypes)

sunshine_question2 = sun.astype({'Month1':'float64',
                 'Month2':'float64',
                 'Month3':'float64',
                 'Month4':'float64',
                 'Month5':'float64',
                'Month6':'float64',
                'Month7':'float64',
                'Month8':'float64',
                'Month9':'float64',
                'Month10':'float64',
                'Month11':'float64',
                'Month12':'float64',
                'Total':'float64'})

print("\nNew Data Types: \n", sunshine_question2.dtypes)

Original Data Types: 
 Location    object
City        object
Month1      object
Month2      object
Month3      object
Month4      object
Month5      object
Month6      object
Month7      object
Month8      object
Month9      object
Month10     object
Month11     object
Month12     object
Total       object
dtype: object

New Data Types: 
 Location     object
City         object
Month1      float64
Month2      float64
Month3      float64
Month4      float64
Month5      float64
Month6      float64
Month7      float64
Month8      float64
Month9      float64
Month10     float64
Month11     float64
Month12     float64
Total       float64
dtype: object


## 3. Fix missing and invalid values in data.

#### Identify missing rows

In [10]:
sunshine.isnull().any(axis = 1).sum()

6

In [11]:
nan_rows  = sunshine[sunshine.isnull().any(axis=1)]

nan_rows

Country       City    Jan    Feb    Mar    Apr    May    Jun    Jul  \
3        NaN    Algiers    NaN    NaN  202.0  258.0  319.0  318.0  350.0   
16       NaN   Canberra  294.5  254.3  251.1  219.0  186.0  156.0  179.8   
25   Belarus        NaN    NaN    NaN    NaN  185.0  270.0    NaN    NaN   
39    Brazil  Fortaleza  222.0  181.2  157.0  147.1  201.3  215.5  259.4   
46  Bulgaria      Sofia   88.0  114.0    NaN  182.0  230.0  258.0  302.0   
53  Cameroon    Yaoundé  177.0  172.0  149.0  153.0  158.0  112.0   85.0   

      Aug    Sep    Oct    Nov    Dec  Yearly  
3   319.0  237.0  229.0  165.0  136.0  2847.0  
16  217.0  231.0  266.6  267.0  291.4     NaN  
25  251.0  154.0  103.0   39.0   28.0     NaN  
39  288.3  284.4    NaN  288.5  265.4  2804.2  
46  288.0  220.0  164.0  106.0   66.0  2177.0  
53   83.0    NaN  120.0  159.0  179.0  1644.0

#### Drop rows with too much missing data

Row index 25 has many missing data points that are too difficult or impossible to identify. For example, I do not know the city and have no way to replace it. I am also missing the total so I can't really replace the missing months using the Yearly column. This row can be dropped.

In [12]:
sunshine.drop(25, axis = 0, inplace = True)

nan_rows  = sunshine[sunshine.isnull().any(axis=1)]

print(sunshine.isnull().any(axis = 1).sum())
nan_rows

5


Country       City    Jan    Feb    Mar    Apr    May    Jun    Jul  \
3        NaN    Algiers    NaN    NaN  202.0  258.0  319.0  318.0  350.0   
16       NaN   Canberra  294.5  254.3  251.1  219.0  186.0  156.0  179.8   
39    Brazil  Fortaleza  222.0  181.2  157.0  147.1  201.3  215.5  259.4   
46  Bulgaria      Sofia   88.0  114.0    NaN  182.0  230.0  258.0  302.0   
53  Cameroon    Yaoundé  177.0  172.0  149.0  153.0  158.0  112.0   85.0   

      Aug    Sep    Oct    Nov    Dec  Yearly  
3   319.0  237.0  229.0  165.0  136.0  2847.0  
16  217.0  231.0  266.6  267.0  291.4     NaN  
39  288.3  284.4    NaN  288.5  265.4  2804.2  
46  288.0  220.0  164.0  106.0   66.0  2177.0  
53   83.0    NaN  120.0  159.0  179.0  1644.0

#### Use deductive reasoning to replace NaN values 

The remaining missing values can be replaced using deductive reasoning:
* Countries can be identified via the city
* Single missing month values can be found by subtracting all other months from Yearly
* Double missing month values can be found  by subtracting all other months from Yearly and split the difference between the two
* Yearly can be found by adding all other months together

As shown below all NaN values have been accounted for.

In [13]:
### Country column replacements ###

# Algiers is the capital of Algeria
sunshine.loc[3,"Country"] = 'Algeria'

# Canberra is the capital of Australia
sunshine.loc[16,"Country"] = 'Australia'


### Month column replacements ###

# Get month totals
month = ['Jan','Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov','Dec']
month_sum = nan_rows[month].sum(numeric_only = True, axis = 1)

# Subtract month total from Yearly
diff = nan_rows['Yearly'] - month_sum

# find number of months with missing values
num_nan = nan_rows[month].isnull().sum(axis=1)

# Now we have enough information to replace NaN values

sunshine.loc[3,"Jan"] = diff[3] / num_nan[3]
sunshine.loc[3,"Feb"] = diff[3] / num_nan[3]
sunshine.loc[16,"Yearly"] = month_sum[16]
sunshine.loc[39,"Oct"] = diff[39] / num_nan[39]
sunshine.loc[46,"Mar"] = diff[46] / num_nan[46]
sunshine.loc[53,"Sep"] = diff[53] / num_nan[53]

nan_rows  = sunshine[sunshine.isnull().any(axis=1)]

print(sunshine.isnull().any(axis = 1).sum())
nan_rows

0


Empty DataFrame
Columns: [Country, City, Jan, Feb, Mar, Apr, May, Jun, Jul, Aug, Sep, Oct, Nov, Dec, Yearly]
Index: []

## 4. Create new columns based on existing columns or calculations.

I was curious to see the amount of sun each city recieved on a daily basis, on average. To create these variables, I took the value for each month and divided it by the appropriate number of days. See function below.

In [14]:
pd.options.mode.chained_assignment = None  # default='warn'

sunshine_non_numeric = sunshine[['Country','City']] 

def get_daily_sunshine(df):
    
    # Get column names
    columns = list(sunshine)
    
    for i in columns:
        if 'Jan' in  i:
            df['Jan_daily_sunshine'] = df['Jan'] / 31
        elif 'Feb' in  i:
            df['Feb_daily_sunshine'] = df['Feb'] / 28
        elif 'Mar' in  i:
            df['Mar_daily_sunshine'] = df['Mar'] / 31
        elif 'Apr' in  i:
            df['Apr_daily_sunshine'] = df['Apr'] / 30
        elif 'May' in  i:
            df['May_daily_sunshine'] = df['May'] / 31
        elif 'Jun' in  i:
            df['Jun_daily_sunshine'] = df['Jun'] / 30
        elif 'Jul' in  i:
            df['Jul_daily_sunshine'] = df['Jul'] / 31
        elif 'Aug' in  i:
            df['Aug_daily_sunshine'] = df['Aug'] / 31
        elif 'Sep' in  i:
            df['Sep_daily_sunshine'] = df['Sep'] / 30
        elif 'Oct' in  i:
            df['Oct_daily_sunshine'] = df['Oct'] / 31
        elif 'Nov' in  i:
            df['Nov_daily_sunshine'] = df['Nov'] / 30
        elif 'Dec' in  i:
            df['Dec_daily_sunshine'] = df['Dec'] / 31
        elif 'Yearly' in  i:
            df['Yearly_daily_sunshine'] = df['Yearly'] / 365
        
    return df

# function works on numeric data only
numeric_data = ['Jan','Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov','Dec', 'Yearly']
sunshine1 = get_daily_sunshine(sunshine[numeric_data])

# column bind non numeric and numeric data together 
sunshine = pd.concat([sunshine_non_numeric, sunshine1], axis=1, ignore_index=False)

sunshine.head()

Country         City    Jan    Feb    Mar    Apr    May    Jun    Jul  \
1  Afghanistan        Kabul  177.2  178.6  204.5  232.5  310.3  353.4  356.8   
2      Albania       Tirana  124.0  125.0  165.0  191.0  263.0  298.0  354.0   
3      Algeria      Algiers  157.0  157.0  202.0  258.0  319.0  318.0  350.0   
4      Algeria  Tamanrasset  297.6  275.5  322.4  327.0  328.6  306.0  356.5   
5       Angola       Luanda  219.0  208.0  213.0  199.0  233.0  223.0  175.0   

     Aug  ...  Apr_daily_sunshine  May_daily_sunshine  Jun_daily_sunshine  \
1  339.7  ...            7.750000           10.009677           11.780000   
2  327.0  ...            6.366667            8.483871            9.933333   
3  319.0  ...            8.600000           10.290323           10.600000   
4  331.7  ...           10.900000           10.600000           10.200000   
5  150.0  ...            6.633333            7.516129            7.433333   

   Jul_daily_sunshine  Aug_daily_sunshine  Sep_daily_sunshine  \
1           11.509677           10.958065           10.130000   
2           11.419355           10.548387            8.800000   
3           11.290323           10.290323            7.900000   
4           11.500000           10.700000            9.600000   
5            5.645161            4.838710            4.833333   

   Oct_daily_sunshine  Nov_daily_sunshine  Dec_daily_sunshine  \
1            9.116129            8.440000            5.883871   
2            7.032258            4.233333            2.838710   
3            7.387097            5.500000            4.387097   
4           10.000000            9.500000            8.800000   
5            5.290323            6.633333            6.838710   

   Yearly_daily_sunshine  
1               8.698904  
2               6.969863  
3               7.800000  
4              10.098630  
5               6.413699  

[5 rows x 28 columns]

## 5. Drop column(s) from your dataset.

For the sake of demonstration, I will copy the dataframe first and then drop columns as I think all of the variables are valuable for further analysis. 

In [15]:
# create copy
sunshine_copy = sunshine.copy()

# drop columns
sunshine_copy = sunshine_copy.drop(columns = ['Jan',
                             'Feb',
                             'Mar',
                             'Apr',
                             'May',
                             'Jun',
                             'Jul',
                             'Aug',
                             'Sep',
                             'Oct',
                             'Nov',
                             'Dec',
                             'Yearly'])

sunshine_copy.head()

Country         City  Jan_daily_sunshine  Feb_daily_sunshine  \
1  Afghanistan        Kabul            5.716129            6.378571   
2      Albania       Tirana            4.000000            4.464286   
3      Algeria      Algiers            5.064516            5.607143   
4      Algeria  Tamanrasset            9.600000            9.839286   
5       Angola       Luanda            7.064516            7.428571   

   Mar_daily_sunshine  Apr_daily_sunshine  May_daily_sunshine  \
1            6.596774            7.750000           10.009677   
2            5.322581            6.366667            8.483871   
3            6.516129            8.600000           10.290323   
4           10.400000           10.900000           10.600000   
5            6.870968            6.633333            7.516129   

   Jun_daily_sunshine  Jul_daily_sunshine  Aug_daily_sunshine  \
1           11.780000           11.509677           10.958065   
2            9.933333           11.419355           10.548387   
3           10.600000           11.290323           10.290323   
4           10.200000           11.500000           10.700000   
5            7.433333            5.645161            4.838710   

   Sep_daily_sunshine  Oct_daily_sunshine  Nov_daily_sunshine  \
1           10.130000            9.116129            8.440000   
2            8.800000            7.032258            4.233333   
3            7.900000            7.387097            5.500000   
4            9.600000           10.000000            9.500000   
5            4.833333            5.290323            6.633333   

   Dec_daily_sunshine  Yearly_daily_sunshine  
1            5.883871               8.698904  
2            2.838710               6.969863  
3            4.387097               7.800000  
4            8.800000              10.098630  
5            6.838710               6.413699

## 6. Drop a row(s) from your dataset. (drop duplicates)

There are duplicate rows that need to be dealt with in this dataset. The `.drop_duplicates` function will do the trick, keeping only the first occurance of a duplicated row.

In [16]:
duplicate = sunshine[sunshine.duplicated()]
 
print("Duplicate Rows :")
 
# Print the resultant Dataframe
duplicate

Duplicate Rows :


Country              City    Jan    Feb    Mar    Apr    May    Jun  \
378   Vietnam            Da Lat  255.0  234.0  255.0  202.0  190.0  147.0   
379   Vietnam           Da Nang  139.0  145.0  188.0  209.0  246.0  239.0   
380   Vietnam             Hanoi   74.0   47.0   47.0   90.0  183.0  172.0   
381   Vietnam  Ho Chi Minh City  245.0  246.0  272.0  239.0  195.0  171.0   
385    Zambia       Livingstone  213.9  196.0  251.1  273.0  303.8  288.0   
386    Zambia            Lusaka  176.7  168.0  220.1  246.0  275.9  270.0   
387    Zambia             Ndola  151.9  142.8  192.2  243.0  279.0  276.0   
390  Zimbabwe            Harare  217.0  190.4  232.5  249.0  269.7  264.0   
391  Zimbabwe            Harare  217.0  190.4  232.5  249.0  269.7  264.0   
392  Zimbabwe            Harare  217.0  190.4  232.5  249.0  269.7  264.0   
393  Zimbabwe            Harare  217.0  190.4  232.5  249.0  269.7  264.0   
394  Zimbabwe            Harare  217.0  190.4  232.5  249.0  269.7  264.0   

       Jul    Aug  ...  Apr_daily_sunshine  May_daily_sunshine  \
378  157.0  136.0  ...            6.733333            6.129032   
379  253.0  218.0  ...            6.966667            7.935484   
380  195.0  174.0  ...            3.000000            5.903226   
381  180.0  172.0  ...            7.966667            6.290323   
385  310.0  319.3  ...            9.100000            9.800000   
386  294.5  303.8  ...            8.200000            8.900000   
387  297.6  297.6  ...            8.100000            9.000000   
390  279.0  300.7  ...            8.300000            8.700000   
391  279.0  300.7  ...            8.300000            8.700000   
392  279.0  300.7  ...            8.300000            8.700000   
393  279.0  300.7  ...            8.300000            8.700000   
394  279.0  300.7  ...            8.300000            8.700000   

     Jun_daily_sunshine  Jul_daily_sunshine  Aug_daily_sunshine  \
378            4.900000            5.064516            4.387097   
379            7.966667            8.161290            7.032258   
380            5.733333            6.290323            5.612903   
381            5.700000            5.806452            5.548387   
385            9.600000           10.000000           10.300000   
386            9.000000            9.500000            9.800000   
387            9.200000            9.600000            9.600000   
390            8.800000            9.000000            9.700000   
391            8.800000            9.000000            9.700000   
392            8.800000            9.000000            9.700000   
393            8.800000            9.000000            9.700000   
394            8.800000            9.000000            9.700000   

     Sep_daily_sunshine  Oct_daily_sunshine  Nov_daily_sunshine  \
378            4.433333            4.516129            5.733333   
379            5.866667            4.677419            4.000000   
380            5.866667            5.387097            4.566667   
381            5.400000            5.870968            6.666667   
385            9.900000            9.000000            7.600000   
386            9.700000            8.800000            7.800000   
387            9.300000            8.700000            6.900000   
390            9.800000            9.200000            7.700000   
391            9.800000            9.200000            7.700000   
392            9.800000            9.200000            7.700000   
393            9.800000            9.200000            7.700000   
394            9.800000            9.200000            7.700000   

     Dec_daily_sunshine  Yearly_daily_sunshine  
378            6.935484               6.131507  
379            3.322581               5.978082  
380            4.000000               4.342466  
381            7.290323               6.819178  
385            6.700000               8.676164  
386            5.900000               8.043014  
387            5.100000               7.654521  
390      

In [17]:
sunshine = sunshine.drop_duplicates(keep = 'first')

duplicate = sunshine[sunshine.duplicated()]
 
print("Duplicate Rows :")
 
# Print the resultant Dataframe
duplicate

Duplicate Rows :


Empty DataFrame
Columns: [Country, City, Jan, Feb, Mar, Apr, May, Jun, Jul, Aug, Sep, Oct, Nov, Dec, Yearly, Jan_daily_sunshine, Feb_daily_sunshine, Mar_daily_sunshine, Apr_daily_sunshine, May_daily_sunshine, Jun_daily_sunshine, Jul_daily_sunshine, Aug_daily_sunshine, Sep_daily_sunshine, Oct_daily_sunshine, Nov_daily_sunshine, Dec_daily_sunshine, Yearly_daily_sunshine]
Index: []

[0 rows x 28 columns]

## 7. Sort your data based on multiple variables.

The sort values functin can be used to sort multiple variables. The ascending argument can be used to fine tune the sort function. Here I am sorting by ascending Yearly hours of sunlight (smallest to largest) and by descending Country name (z to a).  

In [18]:
sunshine.sort_values(['Yearly', 'Country'], ascending=[True, False])

Country          City    Jan    Feb    Mar    Apr    May  \
120      Faroe Islands      Tórshavn   14.0   36.0   71.0  106.0  124.0   
248             Russia        Dikson    0.0   22.6  127.1  237.0  189.1   
113  Equatorial Guinea        Malabo  148.8  152.5  108.5  120.0  117.8   
229             Norway        Bergen   19.0   56.0   94.0  147.0  186.0   
240               Peru          Lima  179.1  169.0  139.2  184.0  116.4   
..                 ...           ...    ...    ...    ...    ...    ...   
350      United States       Phoenix  256.0  257.2  318.4  353.6  401.0   
68               Chile        Calama  353.4  305.1  319.3  312.0  306.9   
108              Egypt  Dakhla Oasis  294.5  279.7  316.2  315.0  356.5   
110              Egypt    Marsa Alam  279.0  283.0  310.0  330.0  372.0   
368      United States          Yuma  268.4  270.8  335.5  365.5  407.4   

       Jun    Jul    Aug  ...  Apr_daily_sunshine  May_daily_sunshine  \
120  125.0  111.0   98.0  ...            3.533333            4.000000   
248  141.0  223.2  139.5  ...            7.900000            6.100000   
113   69.0   46.5   58.9  ...            4.000000            3.800000   
229  189.0  167.0  144.0  ...            4.900000            6.000000   
240   50.6   28.6   32.3  ...            6.133333            3.754839   
..     ...    ...    ...  ...                 ...                 ...   
350  407.8  378.5  360.8  ...           11.786667           12.935484   
68   297.0  310.0  316.2  ...           10.400000            9.900000   
108  366.0  384.4  375.1  ...           10.500000           11.500000   
110  390.0  403.0  372.0  ...           11.000000           12.000000   
368  415.4  392.6  375.6  ...           12.183333           13.141935   

     Jun_daily_sunshine  Jul_daily_sunshine  Aug_daily_sunshine  \
120            4.166667            3.580645            3.161290   
248            4.700000            7.200000            4.500000   
113            2.300000            1.500000            1.900000   
229            6.300000            5.387097            4.645161   
240            1.686667            0.922581            1.041935   
..                  ...                 ...                 ...   
350           13.593333           12.209677           11.638710   
68             9.900000           10.000000           10.200000   
108           12.200000           12.400000           12.100000   
110           13.000000           13.000000           12.000000   
368           13.846667           12.664516           12.116129   

     Sep_daily_sunshine  Oct_daily_sunshine  Nov_daily_sunshine  \
120            2.666667            1.580645            0.666667   
248            2.000000            0.800000            0.000000   
113            1.600000            2.200000            3.300000   
229            2.866667            1.935484            0.900000   
240            1.243333            2.106452            2.966667   
..                  ...                 ...                 ...   
350           10.953333            9.964516            8.533333   
68            10.700000           11.500000           12.100000   
108           11.200000           10.600000           10.000000   
110           11.000000           10.000000           10.000000   
368           11.390000           10.309677            9.003333   

     Dec_daily_sunshine  Yearly_daily_sunshine  
120            0.193548               2.301370  
248            0.000000               3.189863  
113            4.500000               3.223836  
229            0.387097               3.252055  
240            4.490323               3.369863  
..                  ...                    ...  
350            7.896774              10.607123  
68            11.800000              10.756712  
108            9.400000              10.803836  
110            9.000000              10.843836  
368            8.151613              11.000822  

[381 rows x 28 columns]

## 8. Filter your data based on some condition.

I decided to see if there were multiple places that experienced 0 sunlight hours in a month. Based on the filtered dataframe below, Norway and Russia both experienced a month with no sunlight.

In [19]:
filter_df = sunshine[(sunshine['Jan'] == 0)  | (sunshine['Jan'] == 0) |
                     (sunshine['Mar'] == 0)  | (sunshine['Apr'] == 0) |
                     (sunshine['May'] == 0)  | (sunshine['Jun'] == 0) |
                     (sunshine['Jul'] == 0)  | (sunshine['Aug'] == 0) |
                     (sunshine['Sep'] == 0)  | (sunshine['Oct'] == 0) |
                     (sunshine['Nov'] == 0)  | (sunshine['Dec'] == 0)]

filter_df

Country    City  Jan   Feb    Mar    Apr    May    Jun    Jul    Aug  ...  \
231  Norway  Tromsø  3.0  32.0  112.0  160.0  218.0  221.0  205.0  167.0  ...   
248  Russia  Dikson  0.0  22.6  127.1  237.0  189.1  141.0  223.2  139.5  ...   

     Apr_daily_sunshine  May_daily_sunshine  Jun_daily_sunshine  \
231            5.333333            7.032258            7.366667   
248            7.900000            6.100000            4.700000   

     Jul_daily_sunshine  Aug_daily_sunshine  Sep_daily_sunshine  \
231            6.612903            5.387097            3.066667   
248            7.200000            4.500000            2.000000   

     Oct_daily_sunshine  Nov_daily_sunshine  Dec_daily_sunshine  \
231            1.580645                 0.2                 0.0   
248            0.800000                 0.0                 0.0   

     Yearly_daily_sunshine  
231               3.465753  
248               3.189863  

[2 rows x 28 columns]

## 9. Convert all the string values to upper or lower cases in one column.

The `.upper()` or `.lower()` functions can convert strings to its respective case type. For my example, I will use `.upper()` to convert the Country column to upper case. It is important to convert the series to a string first before calling the `.upper` function.

In [20]:
sunshine['Country'] = sunshine['Country'].str.upper()

sunshine.head()

Country         City    Jan    Feb    Mar    Apr    May    Jun    Jul  \
1  AFGHANISTAN        Kabul  177.2  178.6  204.5  232.5  310.3  353.4  356.8   
2      ALBANIA       Tirana  124.0  125.0  165.0  191.0  263.0  298.0  354.0   
3      ALGERIA      Algiers  157.0  157.0  202.0  258.0  319.0  318.0  350.0   
4      ALGERIA  Tamanrasset  297.6  275.5  322.4  327.0  328.6  306.0  356.5   
5       ANGOLA       Luanda  219.0  208.0  213.0  199.0  233.0  223.0  175.0   

     Aug  ...  Apr_daily_sunshine  May_daily_sunshine  Jun_daily_sunshine  \
1  339.7  ...            7.750000           10.009677           11.780000   
2  327.0  ...            6.366667            8.483871            9.933333   
3  319.0  ...            8.600000           10.290323           10.600000   
4  331.7  ...           10.900000           10.600000           10.200000   
5  150.0  ...            6.633333            7.516129            7.433333   

   Jul_daily_sunshine  Aug_daily_sunshine  Sep_daily_sunshine  \
1           11.509677           10.958065           10.130000   
2           11.419355           10.548387            8.800000   
3           11.290323           10.290323            7.900000   
4           11.500000           10.700000            9.600000   
5            5.645161            4.838710            4.833333   

   Oct_daily_sunshine  Nov_daily_sunshine  Dec_daily_sunshine  \
1            9.116129            8.440000            5.883871   
2            7.032258            4.233333            2.838710   
3            7.387097            5.500000            4.387097   
4           10.000000            9.500000            8.800000   
5            5.290323            6.633333            6.838710   

   Yearly_daily_sunshine  
1               8.698904  
2               6.969863  
3               7.800000  
4              10.098630  
5               6.413699  

[5 rows x 28 columns]

## 10. Check whether numeric values are present in a given column of your dataframe.

Pandas has functions to check whether numeric values are present in a given column (`.str.isnumeric()`, `.str.isdigit()`, and `.str.isdecimal()`). The functions work great if there are no decimal points or commas present in the string. The Jan column needed to have the decimals removed to recieve the proper results. I ran it on both a column with non numeric and one with numeric data to show the different results.

In [21]:
sunshine2 = sunshine.copy()

In [22]:
print("Number of numeric values in City: ", sunshine2['City'].str.isdigit().sum())
sunshine2['City'].str.isdigit()

Number of numeric values in City:  0


1      False
2      False
3      False
4      False
5      False
       ...  
382    False
383    False
384    False
388    False
389    False
Name: City, Length: 381, dtype: bool

In [23]:
jan = sunshine2['Jan'].astype('str')
jan = jan.str.replace('.','', regex = True)

print("Number of numeric values in Jan: ", jan.str.isdigit().sum())
jan.str.isdigit()

Number of numeric values in Jan:  381


1      True
2      True
3      True
4      True
5      True
       ... 
382    True
383    True
384    True
388    True
389    True
Name: Jan, Length: 381, dtype: bool

## 11. Group your dataset by one column, and get the mean, min, and max values by group.

### Groupby()

The groupby function will combine all cities in a particular country into a singular groups (Country). 

In [24]:
sunshine.groupby(by='Country').describe()

Jan                                                          \
            count        mean        std    min      25%     50%      75%   
Country                                                                     
AFGHANISTAN   1.0  177.200000        NaN  177.2  177.200  177.20  177.200   
ALBANIA       1.0  124.000000        NaN  124.0  124.000  124.00  124.000   
ALGERIA       2.0  227.300000  99.419213  157.0  192.150  227.30  262.450   
ANGOLA        1.0  219.000000        NaN  219.0  219.000  219.00  219.000   
ARGENTINA     6.0  238.250000  53.943331  164.3  197.925  254.20  273.575   
...           ...         ...        ...    ...      ...     ...      ...   
UZBEKISTAN    1.0  117.800000        NaN  117.8  117.800  117.80  117.800   
VENEZUELA     3.0  259.133333  36.593078  229.4  238.700  248.00  274.000   
VIETNAM       4.0  178.250000  87.091427   74.0  122.750  192.00  247.500   
ZAMBIA        3.0  180.833333  31.205982  151.9  164.300  176.70  195.300   
ZIMBABWE      2.0  230.950000  19.728279  217.0  223.975  230.95  237.925   

                     Feb              ... Dec_daily_sunshine            \
               max count        mean  ...                75%       max   
Country                               ...                                
AFGHANISTAN  177.2   1.0  178.600000  ...           5.883871  5.883871   
ALBANIA      124.0   1.0  125.000000  ...           2.838710  2.838710   
ALGERIA      297.6   2.0  216.250000  ...           7.696774  8.800000   
ANGOLA       219.0   1.0  208.000000  ...           6.838710  6.838710   
ARGENTINA    297.6   6.0  207.066667  ...           8.475000  9.200000   
...            ...   ...         ...  ...                ...       ...   
UZBEKISTAN   117.8   1.0  127.100000  ...           3.400000  3.400000   
VENEZUELA    300.0   3.0  243.900000  ...           8.353226  8.806452   
VIETNAM      255.0   4.0  168.000000  ...           7.024194  7.290323   
ZAMBIA       213.9   3.0  168.933333  ...           6.300000  6.700000   
ZIMBABWE     244.9   2.0  201.600000  ...           7.075000  7.300000   

            Yearly_daily_sunshine                                          \
                            count      mean       std       min       25%   
Country                                                                     
AFGHANISTAN                   1.0  8.698904       NaN  8.698904  8.698904   
ALBANIA                       1.0  6.969863       NaN  6.969863  6.969863   
ALGERIA                       2.0  8.949315  1.625377  7.800000  8.374658   
ANGOLA                        1.0  6.413699       NaN  6.413699  6.413699   
ARGENTINA                     6.0  6.083014  1.576948  3.510137  5.317123   
...                           ...       ...       ...       ...       ...   
UZBEKISTAN                    1.0  7.736712       NaN  7.736712  7.736712   
VENEZUELA                     3.0  7.932329  1.064795  6.868219  7.399589   
VIETNAM                       4.0  5.817808  1.049360  4.342466  5.569178   
ZAMBIA                        3.0  8.124566  0.515681  7.654521  7.848767   
ZIMBABWE                      2.0  8.398356  0.211163  8.249041  8.323699   

                                            
                  50%       75%        max  
Country                                     
AFGHANISTAN  8.698904  8.698904   8.698904  
ALBANIA      6.969863  6.969863   6.969863  
ALGERIA      8.949315  9.523973  10.098630  
ANGOLA       6.413699  6.413699   6.413699  
ARGENTINA    6.586849  6.936233   7.867671  
...               ...       ...        ...  
UZBEKISTAN   7.736712  7.736712   7.736712  
VENEZUELA    7.930959  8.464384   8.997808  
VIETNAM      6.054795  6.303425   6.819178  
ZAMBIA       8.043014  8.359589   8.676164  
ZIMBABWE     8.398356  8.473014   8.547671  

[139 rows x 208 columns]

### .agg() and .apply()

The agg function aggregates the data using one or more operations. Below shows the data grouped by country and aggrated by mean.

In [25]:
sunshine.groupby(by='Country')['Yearly_daily_sunshine'].agg('mean')

Country
AFGHANISTAN    8.698904
ALBANIA        6.969863
ALGERIA        8.949315
ANGOLA         6.413699
ARGENTINA      6.083014
                 ...   
UZBEKISTAN     7.736712
VENEZUELA      7.932329
VIETNAM        5.817808
ZAMBIA         8.124566
ZIMBABWE       8.398356
Name: Yearly_daily_sunshine, Length: 139, dtype: float64

The .apply function is used to apply functions to columns in the dataframe. Below I created a column to assign if a city recieves a large amount of sunlight. Most cities recieved below 8 hours of sunlight as a yearly daily average.

In [26]:
def assign_level(column):
    if column < 8:
        return 'low sunlight'
    elif column >= 8 and column <= 16:
        return 'avg. sunlight'
    elif column > 16:
        return 'high sunlight'
    
sunshine['sunshine_level'] = sunshine['Yearly_daily_sunshine'].apply(assign_level)

sunshine['sunshine_level']

1      avg. sunlight
2       low sunlight
3       low sunlight
4      avg. sunlight
5       low sunlight
           ...      
382    avg. sunlight
383    avg. sunlight
384     low sunlight
388    avg. sunlight
389    avg. sunlight
Name: sunshine_level, Length: 381, dtype: object

In [27]:
sunshine.groupby(by = 'sunshine_level')['sunshine_level'].agg('count')

sunshine_level
avg. sunlight    109
low sunlight     272
Name: sunshine_level, dtype: int64

In [28]:
sunshine.groupby(by = 'Country')['Country'].agg('count').sort_values(ascending=False)

Country
UNITED STATES    54
CHINA            15
BRAZIL           11
AUSTRALIA         9
CANADA            9
                 ..
ICELAND           1
HUNGARY           1
HONDURAS          1
GUYANA            1
LATVIA            1
Name: Country, Length: 139, dtype: int64

## 12. Group your dataset by two columns and then sort the aggregated results within the groups.

Grouping by two columns requires that the sorted values function take a tuple as the olumn becomes multilevel.

In [29]:
q12 = sunshine.groupby(['Country', 'sunshine_level']).agg({'Jan': ['mean', 'min', 'max'],
                                                    'Feb': ['mean', 'min', 'max'],
                                                    'Mar': ['mean', 'min', 'max'],
                                                    'Apr': ['mean', 'min', 'max'],
                                                    'May': ['mean', 'min', 'max'],
                                                    'Jun': ['mean', 'min', 'max'],
                                                    'Jul': ['mean', 'min', 'max'],
                                                    'Aug': ['mean', 'min', 'max'],
                                                    'Sep': ['mean', 'min', 'max'],
                                                    'Oct': ['mean', 'min', 'max'],
                                                    'Nov': ['mean', 'min', 'max'],
                                                    'Dec': ['mean', 'min', 'max']})
# need to provide tuple as it is a multi level column
q12.sort_values([('Jan', 'mean')], ascending=[True])

Jan                       Feb         \
                                    mean    min    max        mean    min   
Country       sunshine_level                                                
FAROE ISLANDS low sunlight     14.000000   14.0   14.0   36.000000   36.0   
ICELAND       low sunlight     20.000000   20.0   20.0   60.000000   60.0   
NORWAY        low sunlight     20.666667    3.0   40.0   54.666667   32.0   
LATVIA        low sunlight     31.000000   31.0   31.0   62.000000   62.0   
UKRAINE       low sunlight     31.000000   31.0   31.0   57.000000   57.0   
...                                  ...    ...    ...         ...    ...   
ERITREA       avg. sunlight   304.700000  304.7  304.7  289.100000  289.1   
CHAD          avg. sunlight   306.900000  297.6  316.2  284.200000  277.2   
CHILE         low sunlight    320.333333  257.3  362.7  269.700000  228.8   
NAMIBIA       avg. sunlight   320.500000  288.0  353.0  277.000000  254.0   
CHILE         avg. sunlight   336.350000  319.3  353.4  298.050000  291.0   

                                            Mar                       Apr  \
                                max        mean    min    max        mean   
Country       sunshine_level                                                
FAROE ISLANDS low sunlight     36.0   71.000000   71.0   71.0  106.000000   
ICELAND       low sunlight     60.0  109.000000  109.0  109.0  164.000000   
NORWAY        low sunlight     76.0  110.666667   94.0  126.0  161.666667   
LATVIA        low sunlight     62.0  127.000000  127.0  127.0  183.000000   
UKRAINE       low sunlight     57.0  124.000000  124.0  124.0  180.000000   
...                             ...         ...    ...    ...         ...   
ERITREA       avg. sunlight   289.1  300.700000  300.7  300.7  300.500000   
CHAD          avg. sunlight   291.2  291.400000  282.1  300.7  286.500000   
CHILE         low sunlight    302.3  244.133333  204.6  272.8  171.666667   
NAMIBIA       avg. sunlight   300.0  297.000000  282.0  312.0  289.500000   
CHILE         avg. sunlight   305.1  306.900000  294.5  319.3  282.000000   

                              ...    Sep     Oct                       Nov  \
                              ...    max    mean    min    max        mean   
Country       sunshine_level  ...                                            
FAROE ISLANDS low sunlight    ...   80.0   49.00   49.0   49.0   20.000000   
ICELAND       low sunlight    ...  120.0   93.00   93.0   93.0   41.000000   
NORWAY        low sunlight    ...  144.0   65.00   49.0   86.0   28.000000   
LATVIA        low sunlight    ...  153.0   93.00   93.0   93.0   39.000000   
UKRAINE       low sunlight    ...  210.0  155.00  155.0  155.0   60.000000   
...                           ...    ...     ...    ...    ...         ...   
ERITREA       avg. sunlight   ...  260.4  301.00  301.0  301.0  292.500000   
CHAD          avg. sunlight   ...  261.0  296.05  285.2  306.9  306.000000   
CHILE         low sunlight    ...  197.0  209.70  127.1  254.0  263.666667   
NAMIBIA       avg. sunlight   ...  321.0  331.00  319.0  343.0  322.500000   
CHILE         avg. sunlight   ...  321.0  308.45  260.4  356.5  319.500000   

                                                   Dec                
                                min    max        mean    min    max  
Country       sunshine_level                                          
FAROE ISLANDS low sunlight     20.0   20.0    6.000000    6.0    6.0  
ICELAND       low sunlight     41.0   41.0   22.000000   22.0   22.0  
NORWAY        low sunlight      6.0   51.0   15.666667    0.0   35.0  
LATVIA        low sunlight     39.0   39.0   22.000000   22.0   22.0  
UKRAINE       low sunlight     60.0   60.0   31.000000   31.0   31.0  
...                             ...    ...         ...    ...    ...  
ERITREA       avg. sunlight   292.5  292.5  299.100000  299.1  299.1  
CHAD          avg. sunlight   300.0  312.0  311.550000  303.8  319.3  
CHILE

# Conclusions  

I noticed that the some locations only had one city accounted for, while other places, like the United States and China had many cities. I tested to see if any location had an daily yearly average of greater than 16 hours of sunlight. I found that no location experiences that much sunlight year round. I also found that most cities average less than 8 hours of sunlight daily across an entire year (272/381 = 71%). With more time I would look into the seasons and see how different locations differ from month to month.